### PR TITLE
Ordered Map implementation

### PR DESCRIPTION
**Pull request by yotors**
## Description

This PR implements an ordered map data structure along with several associated functions. The ordered map maintains the order in which key-value pairs are added. The following functions are included:

* `map.add`: Adds a new key-value pair to the map. Returns an error if the key already exists. Maintains sorted order based on keys.
* `map.getByKey`: Retrieves the value associated with a given key. Returns an error if the key is not found.
* `map.contains`: Checks if the map contains a given key.
* `map.valueAtKey`: Checks if the map contains a key with a specific value.
* `map.replace`: Replaces the value associated with a given key. Returns an error if the key is not found.
* `map.remove`: Removes the key-value pair associated with a given key. Returns an error if the key is not found.
* `map.length`: returns the number of elements in the map.

## Motivation and Context

This implementation provides a basic ordered map functionality that other parts of the project can leverage—this helps to access a value using a key.

## How Has This Been Tested?

The ordered map implementation has been thoroughly tested using tests covering all functions and edge cases (adding/removing at beginning/middle/end, duplicate keys, etc.). All tests passed.

A specific example test verifies that `map.add` correctly handles duplicate keys and keeps the existing order:

```metta
! (assertEqual (map.add 1 'a (Consmap (1 'b) Nilmap)) (Error 1 "already exists"))
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: utilities/map.metta [hunk 1]
```diff
@@ -0,0 +1,49 @@
+;; ordered map Data Structure
+(: map (-> $a Type))
+(: Nilmap  (Error (Key: $key not found))(map $a))
```
[the remark this answers is not in the document]
Comment (Yagth) [reply utilities/map.metta:3]: And also change the names using PascalCase. 
```suggestion
(: NilMap (Map $k $v))
(: ConsMap ...)
```

**Review by Yagth — CHANGES_REQUESTED** [review]

**Review by Yagth — CHANGES_REQUESTED** [review]
Your test file should be rewritten as `map-test.metta` otherwise, it won't be picked by our testing script.

File: utilities/map.metta [hunk 2]
```diff
@@ -0,0 +1,49 @@
+;; ordered map Data Structure
+(: map (-> $a Type))
```
[the remark this answers is not in the document]
Comment (yotors) [reply utilities/map.metta:2]: yeah, but both the key and values are in a single expression

File: utilities/map.metta (hunk 1 above)
[the remark this answers is not in the document]
Comment (yotors) [reply utilities/map.metta:3]: oh 

> Why is there an `Error` in this type definition? Shouldn't it be just `(map $a)`.

Oh yeah, it was a complete mistake 💯 :)

File: utilities/map.metta [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+(: Nilmap  (Error (Key: $key not found))(map $a))
+(: Consmap (-> $a (map $a) (map $a)))
+
+;; add an element to an ordered map
+(: map.add (-> $a $b (map $a) (map $a)))
```
[the remark this answers is not in the document]
Comment (yotors) [reply utilities/map.metta:7]: > Rename the function to `insert`. Most of the functions in the other datastructures we have use that name so best to keep it consistent.

ok I will. I saw "add function should be implemented", that was why I made it add.

[the remark this answers is not in the document]
Comment (yotors) [reply utilities/map.metta:7]: > And since the added value is stored like a pair, why not make the function accept a pair too? That would make the usage more consistent.

Yeah, it makes sense to accept as expression 👍

[the remark this answers is not in the document]
Comment (yotors) [reply utilities/map.metta:7]: > One last comment about this function is that most of the `map` datastructures in other languages overwrite the value if the key already exists instead of returning an error. Best if we follow the same logic here to avoid misconcpetion in usage.

Ok. I thought one should use .replace for that, but It would make more sense if I do it as you said

**Review by Yagth — APPROVED** [review]

File: utilities/map.metta (hunk 2 above)
[the remark this answers is not in the document]
Comment (Yagth) [reply utilities/map.metta]: Yeah. It's best if you change it. It's best the key and value types are separate because in different functions we may use them differently.

[the remark this answers is not in the document]
Comment (Yagth) [reply utilities/map.metta]: For example, the `getByKey` function is returning a value, but the type of the function is still a tuple. For example if `$a` is `(String Number)` then the return type of the `getByKey` should have been just `Number` but instead of that the function's return type is now `(String Number)`.

[the remark this answers is not in the document]
Comment (yotors) [reply utilities/map.metta]: Make sense. 👍. Then, what if I use something like `(: Map (-> $k $v Type))`? To keep them in pairs. Then in `getByKey` I would use `(-> $k (Map ($k $v)) $v)`

[the remark this answers is not in the document]
Comment (Yagth) [reply utilities/map.metta]: You don't need that second `-> `before the `$k` but that works

[the remark this answers is not in the document]
Comment (yotors) [reply utilities/map.metta]: k. tnx

**Review by Yagth — APPROVED** [review]

**Comment by Yagth** [comment]
[Ordered map](https://trello.com/c/N5yWK2iB/144-ordered-map)
